pkg/kubelet/volumemanager/reconciler/reconciler.go

```diff
@@ -516,6 +507,16 @@ func (rc *reconciler) reconstructVolume(volume podVolume) (*reconstructedVolume,
 			newMounterErr)
 	}
 
+	// Check existence of mount point for filesystem volume or symbolic link for block volume
+	isExist, checkErr := rc.operationExecutor.CheckVolumeExistenceOperation(volumeSpec, volumeMounter.GetPath(), volumeSpec.Name(), rc.mounter, uniqueVolumeName, volume.podName, pod.UID, attachablePlugin)
+	if checkErr != nil {
+		return nil, checkErr
+	}
+	// If mount or symlink doesn't exist, volume reconstruction should be failed
```

Comment:
@jingxu97 can this check be removed?  This means that the directory still exists and should be cleaned up.

Also this assumes that all volumes use mounts, which is not necessarily true (esp for ephemeral types).

Reply:
For reference, this is the PR that introduced this check.  Reading through the original bug and discussion, I'm not sure how this fix actually addresses the original issue... cc @chakri-nelluri 

https://github.com/kubernetes/kubernetes/pull/51985

Reply:
@msau42 PR #51985 adds an extra check to make sure that the volume is mounted on a reboot and return a failure for volume reconstruction, if it is not. Hope this clarifies the need for check.

Reply:
Ok @jingxu97 and I discussed and I understand now.

If the volume was partially torn down (ie the mount point is gone, but the directory still exists), then we don't want reconstruction to add the volume back to the ASW (because it's actually not mounted) and cause an empty directory to be mounted into the pod.  Let's update the comment here to make that reason more obvious.

However, this brings up some other issues:
* What happens if the volume plugin doesn't actually use mounts? Maybe this isn't an issue for attachable plugins. I can't think of a real attachable plugin that is not mount-based.
* If there was a case of kubelet/node crash in the middle of cleanup, then we may not be able to properly cleanup the directories for CSI because we only try reconstruction once during kubelet initialization, when the CSI plugin may not be registered yet.  This is also more of a corner case and we don't need to consider fixing it urgently for now.

Reply:
> @jingxu97 can this check be removed? This means that the directory still exists and should be cleaned up.
> 
> Also this assumes that all volumes use mounts, which is not necessarily true (esp for ephemeral types).

@msau42 This code is conditional and was only meant for plugins which support attach and needs to be reattached and mounted on a node reboot.

Reply:
> * What happens if the volume plugin doesn't actually use mounts? Maybe this isn't an issue for attachable plugins. I can't think of a real attachable plugin that is not mount-based.

Block devices?

Reply:
> * If there was a case of kubelet/node crash in the middle of cleanup, then we may not be able to properly cleanup the directories for CSI because we only try reconstruction once during kubelet initialization, when the CSI plugin may not be registered yet.  This is also more of a corner case and we don't need to consider fixing it urgently for now.

@msau42 This might show up if we restart Kubelet to recover for any other issues..

Reply:
This CheckVolumeExistance function does handle raw block, although I wonder if it possible to the link to change across reboots? I think this is what the losetup is supposed to protect against. cc @wongma7

Reply:
only kubelet manipulates these links, technically it can change but so can a mount device (we check that a mount point exists only, not if it points to the right thing). (losetup doesn't help here as the failure scenario the losetup code is intended to prevent is something different like: volume at /dev/sda somehow gets attached then another volume gets attached at /dev/sda, silently. The loop device holding onto /dev/sda is supposed to prevent detach from happening at all. If there's a reboot, I don't believe the loop devices persist anyhow.)

Reply:
We should test the reboot behavior for raw block separately to make sure we're not going to accidentally give the wrong device to container if the device names change on a reboot.